### PR TITLE
Fix issue 1163

### LINKED ISSUE #1163: Reference counting bug with NRT+Generators
the cleanup block in the python wrapper for generators removed all NRT reference before the execution of the actual function.

### PR DESCRIPTION
**Pull request by sklam**
Fixes #1163

### CODE REVIEW

**Comment by pitrou** [comment]
Hmm... why can't you handle it the same way CPython references are handled in the object mode version? I'm not sure I understand why you decref the objects *every time* the generator is suspended.

**Comment by sklam** [comment]
My bad.  I was too confused on making a fix quickly that I forgot I was working around something weird.  The nrt references are acquired at each reentry to the generator.  And, they need to be cleaned up.  I am not sure where it acquiring occurred.

**Comment by sklam** [comment]
I think I know what is going on now.

File: numba/generators.py [hunk 1]
```diff
@@ -186,6 +191,40 @@ def get_generator_type(self):
...
+        """
+        Lower the body of the generator's finalizer: decref all live
+        state variables.
+        """
+        if self.context.enable_nrt:
+            resume_index_ptr = cgutils.gep(builder, genptr, 0, 0,
+                                           name='gen.resume_index')
+            resume_index = builder.load(resume_index_ptr)
+
+            # If resume_index is 0, next() was never called
```
Comment (pitrou) [inline numba/generators.py]: Does it mean next() decref's the arguments? Perhaps you could explain why.

File: numba/tests/test_generators.py [hunk 2]
```diff
@@ -214,5 +215,120 @@ def test_return_generator_expr(self):
...
+
+        np.testing.assert_equal(py_ary, c_ary)
+        self.assertEqual(py_res, c_res)
+        # Check reference count
+        self.assertEqual(sys.getrefcount(py_ary),
+                         sys.getrefcount(c_ary))
+
+    def test_nrt_gen0_no_iter(self):
+        """
+        Test cleanup on StopIteration
```
Comment (pitrou) [inline numba/tests/test_generators.py]: The docstring is wrong here :)

**Pull request merged**